java/arcs/core/util/Scheduler.kt

```diff
@@ -0,0 +1,242 @@
+        ) : Task(block)
+    }
+
+    /**
+     * Agenda of [Task]s to be executed by the [Scheduler] on a pass.
+     *
+     * The [Scheduler] should execute all [processors] before executing [listeners] on a
+     * namespace-by-namespace basis.
+     */
+    private data class Agenda(
+        /**
+         * Scheduled [Task.Processor]s.
+         */
+        val processors: List<Task.Processor> = emptyList(),
+        /**
+         * Scheduled [Task.Listener]s, collated by [Task.Listener.namespace].
+         */
+        val listenersByNamespace: ListenersByNamespace = ListenersByNamespace()
+    ) : Iterable<Task> {
+        fun addTasks(tasks: Iterable<Task>): Agenda =
+            tasks.fold(this) { agenda, task -> agenda.addTask(task) }
+
+        fun addTask(task: Task): Agenda = when (task) {
+            is Task.Processor -> addProcessor(task)
+            is Task.Listener -> addListener(task)
+        }
+
+        fun addProcessor(processor: Task.Processor): Agenda =
+            copy(processors = processors + processor)
+
+        fun addListener(listener: Task.Listener): Agenda =
+            copy(listenersByNamespace = listenersByNamespace.addListener(listener))
+
+        fun isEmpty(): Boolean =
+            processors.isEmpty() && listenersByNamespace.isEmpty()
+
+        override fun iterator(): Iterator<Task> = (processors + listenersByNamespace).iterator()
+    }
+
+    /**
+     * [Task.Listener]s, collated by [Task.Listener.namespace].
+     */
+    private data class ListenersByNamespace(
+        val listeners: Map<String, ListenersByName> = emptyMap()
+    ) : Iterable<Task> {
+        fun addListener(listener: Task.Listener): ListenersByNamespace {
+            val listeners = (listeners[listener.namespace] ?: ListenersByName())
+                .addListener(listener)
+
+            return copy(listeners = this.listeners + (listener.namespace to listeners))
+        }
+
+        fun isEmpty(): Boolean = listeners.isEmpty()
+
+        override fun iterator(): Iterator<Task> = listeners.values.flatten().iterator()
+    }
+
+    /**
+     * [Task.Listener]s, collated by [Task.Listener.name].
+     */
+    private data class ListenersByName(
+        val listeners: Map<String, List<Task.Listener>> = emptyMap()
+    ) : Iterable<Task> {
+        fun addListener(listener: Task.Listener): ListenersByName {
+            val resultList = (listeners[listener.name] ?: emptyList()) + listener
+            return copy(listeners = listeners + (listener.name to resultList))
+        }
+
+        override fun iterator(): Iterator<Task> = listeners.values.flatten().iterator()
+    }
+
+    companion object {
+        /**
+         * The default maximum duration each iteration of scheduled tasks' execution is allowed to
+         * take.
+         */
+        const val DEFAULT_AGENDA_PROCESSING_TIMEOUT = 5000L
+
+        /**
+         * The default maximum rate at which iterations of agenda processing are allowed to operate.
```

Comment:
... per second.

Reply:
HZ == per second 😉 

I'll add it though.